stream_test.go

```diff
@@ -371,34 +371,35 @@ func testShutdownAfterClose(t *testing.T, fs FileSystem) {
 
 	wg := sync.WaitGroup{}
 
-	wg.Add(2)
+	wg.Add(3)
 
 	go func() {
-		time.Sleep(50 * time.Millisecond)
-		_, err := ioutil.ReadAll(r)
-		if err != nil {
-			t.Error("Shutdown should allow for any already created readers to finish reading")
-		}
-		r.Close()
+		f.Write([]byte("Hello"))
+		f.Close()
+
+		// This waits for any reads to finish, but prevents new reads as well
+		f.ShutdownWithErr(ErrCanceled)
 		wg.Done()
 	}()
 
-	f.Write([]byte("Hello"))
-	f.Close()
-
-	// This waits for any reads to finish, but prevents new reads as well
-	er := errors.New("shutdown")
-	f.ShutdownWithErr(er)
-
 	go func() {
 		time.Sleep(50 * time.Millisecond)
 		_, err := f.NextReader()
-		if err != er {
+		if err != ErrCanceled {
 			t.Error("Opening new reader after canceling should fail")
 		}
 		wg.Done()
 	}()
 
+	go func() {
+		time.Sleep(50 * time.Millisecond)
+		_, err := ioutil.ReadAll(r)
```

Comment:
You never close this Reader, so ShutdownWithErr blocks forever.

Reply:
Allright, in my implementation I close readers automatically when they read until the end of writer. I guess this design decision is up to you.